File: composition/Scala/composition-pipeline/pipeline/src/test/scala/ResultTests.scala
import DomainModel.Person
import common.ActionPipeline.Result
import org.scalatest.FunSpec

class ResultTests extends FunSpec {
  val NAME_STUB = "John Kidd"
  val EMAIL_ADDRESS_STUB = "jk@test.com"

  def updateNameWithEmail(person: Person): Result[Person] = {
    return Result(data = Person(person.name, EMAIL_ADDRESS_STUB))
  }

  describe("Request") {
    it("should return error message when error provided") {
      // arrange
      val errorStub = "an error"
      val target = Result(errorStub)

      // act
      val actual = target.error

      // assert
      assert(errorStub == actual)
    }

    it ("should return the data") {
      // arrange
      val personStub = Person(name = NAME_STUB)

      // act
      val target = new Result(data = personStub)
      val actual = target.data

      // assert
      assert(personStub == actual)
    }

    it ("should return an updated clone of the data") {
      // arrange
      val personStub = Person(name = NAME_STUB)

      // act
      val actual = updateNameWithEmail(personStub).data

      // assert
      assert(personStub != actual)
    }

    it ("should return an clone of the data with additional field") {
      // arrange
      val personStub = Person(name = NAME_STUB)
      val expected = EMAIL_ADDRESS_STUB

      // act
      val actual = updateNameWithEmail(personStub).data.emailAddress

      // assert
      assert(expected == actual)
    }
  }
}

File: composition/C#/CompositionPipeline.Tests/ResultTests.cs
using NUnit.Framework;
using static NUnit.Framework.Assert;
using CompositionPipeline.Common;

namespace CompositionPipeline.Tests
{
	[TestFixture]
	public class ResultTests
	{
		const string NAME_STUB = "John Kidd";
		const string EMAIL_ADDRESS_STUB = "jim.bob@waltons.com";

		static Result<Person> updateEmailAddress(Person person) {
			return new Result<Person>(data: new Person(name: person.Name, emailAddress: EMAIL_ADDRESS_STUB));
		}

		[Test]
		public void should_return_error_message_when_error_provided() {
			// arrange
			var errorStub = "an error";
			var target = new Result<string>(errorStub);

			// act
			var actual = target.Error;

			// assert
			AreEqual(errorStub, actual);
		}

		[Test]
		public void should_return_the_data() {
			// arrange
			var personStub = new Person(name: NAME_STUB);

			// act
			var target = new Result<Person>(data: personStub);
			var actual = target.Data;

			// assert
			AreEqual(personStub, actual);
		}

		[Test]
		public void should_return_an_updated_clone_of_the_data() {
			// arrange
			var personStub = new Person(name: NAME_STUB);

			// act
			var actual = updateEmailAddress(personStub).Data;

			// assert
			AreNotEqual(personStub, actual);
		}

		[Test]
		public void should_return_an_clone_of_the_data_with_additional_field() {
			// arrange
			var personStub = new Person(name: NAME_STUB);
			var expected = EMAIL_ADDRESS_STUB;

			// act
			var actual = updateEmailAddress(personStub).Data.EmailAddress;

			// assert
			AreEqual(expected, actual);
		}
	}
}
